src/code_builder.rs

```diff
@@ -250,17 +261,43 @@ enum ControlKind {
     Loop,
 }
 
-impl Default for CodeBuilderAllocations {
-    fn default() -> Self {
+impl CodeBuilderAllocations {
+    pub(crate) fn new(module: &Module) -> Self {
+        let mut mutable_globals = HashMap::new();
+        for (i, global) in module
+            .imports
+            .iter()
+            .filter_map(|(_, _, imp)| match imp {
+                Import::Global(g) => Some(g),
+                _ => None,
+            })
+            .chain(module.globals.iter().map(|g| &g.ty))
+            .enumerate()
+        {
+            if global.mutable {
+                mutable_globals
+                    .entry(global.val_type)
+                    .or_insert(Vec::new())
+                    .push(i as u32);
+            }
+        }
+
+        let mut functions = HashMap::new();
+        for (idx, func) in module.funcs() {
+            functions
+                .entry(func.params.clone())
+                .or_insert(Vec::new())
```

Comment:
```suggestion
                .or_default()
```

Reply:
Ah this ends up having issues with type inference because then I'd have to specify that the value of the hash map is a `Vec<T>`, since the `.push(...)` afterwards needs to know what's being pushed on.